todo_app/app.py

```diff
@@ -11,34 +14,88 @@ def create_app():
     config = Config()
     app.config.from_object(config)
 
+    login_manager = LoginManager()
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        github_login = "https://github.com/login/oauth/authorize?client_id=" + config.CLIENT_ID
+        return redirect(github_login)
+
+    @login_manager.user_loader
+    def load_user(user_id):
+        return User(user_id)
+
+    login_manager.init_app(app)
+
     @app.route('/')
+    @login_required
     def index():
         items = get_items(config)
         items = sorted(items, key=lambda item: item.get_ordering_index())
         item_view_model = ViewModel(items)
-        return render_template("index.html", item_view_model=item_view_model)
+        isWriter = current_user.is_anonymous or current_user.role == UserRoles.WRITER
+        return render_template("index.html", item_view_model=item_view_model, isWriter=isWriter)
+
+    @app.route('/login/callback')
+    def user_log_in():
+        code = request.args.get("code")
+
+        data = {
+            "client_id": config.CLIENT_ID,
+            "client_secret": config.CLIENT_SECRET,
+            "code": code
+        }
+
+        access_token_response = requests.post(
+            "https://github.com/login/oauth/access_token", params=data, headers={"Accept": "application/json"})
+
+        access_token_response = access_token_response.json()
+        access_token = access_token_response['access_token']
+
+        user_info_response = requests.get("https://api.github.com/user",
+                                          headers={"Accept": "application/json", "Authorization": "Bearer {}".format(access_token)})
+
+        user_info = user_info_response.json()
+
+        print("logging in as: " + user_info["login"])
```

Comment:
This is definitely helpful in testing, but does have a potential concern in that we'll be logging out usernames which can be easy to identify to individuals. Would outputting an id be a better move?